src/appleseed/renderer/modeling/shadergroup/shader.cpp

```diff
@@ -257,19 +276,60 @@ const char* Shader::get_layer() const
     return get_name();
 }
 
+const char* Shader::get_source_code() const
+{
+    return impl->m_source_code.empty() ? nullptr : impl->m_source_code.c_str();
+}
+
 const ShaderParamContainer& Shader::shader_params() const
 {
     return impl->m_params;
 }
 
+bool Shader::compile_shader(const ShaderCompiler* compiler)
+{
+    // Skip already compiled shaders.
+    if (!impl->m_byte_code.empty())
+        return true;
+
+    if (!impl->m_source_code.empty())
+    {
+        if (!compiler)
+        {
+            RENDERER_LOG_ERROR(
+                "OSL source shader found, but shader compiler is not available.");
+            return false;
+        }
+
+        APIString buffer;
+        const bool success = compiler->compile_buffer(impl->m_source_code.c_str(), buffer);
+
+        if (success)
+            impl->m_byte_code = buffer.c_str();
```

Comment:
Where are shader compilation failure messages printed?

Reply:
We use the OIIOErrorHandler class to log them, same as any other OSL error or warning.